src/components/Editor.tsx

```diff
@@ -1,55 +1,105 @@
 
-const extensions = [
-  StarterKit.configure({
-    bulletList: {
-      keepMarks: true,
-      keepAttributes: false, // TODO : Making this as `false` becase marks are not preserved when I try to preserve attrs, awaiting a bit of help
-    },
-    orderedList: {
-      keepMarks: true,
-      keepAttributes: false, // TODO : Making this as `false` becase marks are not preserved when I try to preserve attrs, awaiting a bit of help
-    },
-  }),
-]
-
-const content = `
-<h2>
-  Hi there,
-</h2>
-<p>
-  this is a <em>basic</em> example of <strong>tiptap</strong>. Sure, there are all kind of basic text styles you’d probably expect from a text editor. But wait until you see the lists:
-</p>
-<ul>
-  <li>
-    That’s a bullet list with one …
-  </li>
-  <li>
-    … or two list items.
-  </li>
-</ul>
-<p>
-  Isn’t that great? And all of that is editable. But wait, there’s more. Let’s try a code block:
-</p>
-<pre><code class="language-css">body {
-display: none;
-}</code></pre>
-<p>
-  I know, I know, this is impressive. It’s only the tip of the iceberg though. Give it a try and click a little bit around. Don’t forget to check the other examples too.
-</p>
-<blockquote>
-  Wow, that’s amazing. Good work, boy! 👏
-  <br />
-  — Mom
-</blockquote>
-`
+const ydoc = new Y.Doc(); // grab from db
+
+const websocketProvider = new HocuspocusProvider({
+  url: "ws://127.0.0.1:80",
+  name: "example-document-room-id",
+  document: ydoc,
+});
+
+const colors = ['#958DF1', '#F98181', '#FBBC88', '#FAF594', '#70CFF8', '#94FADB', '#B9F18D']
+const getRandomElement = (list: string[]) => list[Math.floor(Math.random() * list.length)]
+const getRandomColor = () => {
+  return getRandomElement(colors)
+}
+const getInitialUser = () => {
+  const stored = localStorage.getItem('currentUser');
+  return stored ? JSON.parse(stored) : {
+    name: '',
+    color: getRandomColor(),
+  }
+}
 
 const Editor = () => {
+  const [status, setStatus] = useState('connecting')
+  const [currentUser, setCurrentUser] = useState(getInitialUser)
+
+  const setName = useCallback(() => {
+    const name = (window.prompt('Name') || '').trim().substring(0, 32)
+
+    if (name) {
+      return setCurrentUser({ ...currentUser, name })
+    }
+  }, [currentUser])
+
+  const editor = useEditor({
+    extensions: [
+      StarterKit.configure({
+        history: false,
+        bulletList: {
+          keepMarks: true,
+          keepAttributes: false,
+        },
+        orderedList: {
+          keepMarks: true,
+          keepAttributes: false,
+        },
+      }),
+      CharacterCount.configure({
+        limit: 10000,
+      }),
+      Collaboration.configure({
+        document: ydoc,
+      }),
+      CollaborationCursor.configure({
+        provider: websocketProvider,
+      }),
+    ]
+  })
+
+    useEffect(() => {
+    // Update status changes
+      websocketProvider.on('status', (event: HocuspocusProviderWebsocket) => {
+        setStatus(event.status)
+      })
```

Comment:
I think we need to return a "cleanup" function here, e.g.
```
return () => websocketProvider.off('status', ...etc)
```
or we could move this to the onStatus when it initializes?

Reply:
I changed this to use a config instead of binding an event listener, so I don't think we need to do anything here.

Reply:
edit: I take it back, I found that when applied at the time of initialization, it causes a re-render loop when the websocket can't connect and the whole thing explodes. So I added un-binding. Also, I attempted to set the number of max tries, however it does not seem to work? at least locally.

Reply:
ok one more comment: I removed maxattempts because typescript failure; it's listed in their docs, and it appears to be in a parent type, although I'm not totally sure:

`export type HocuspocusProviderConfiguration = Required<Pick<CompleteHocuspocusProviderConfiguration, 'name'>> & Partial<CompleteHocuspocusProviderConfiguration> & (Required<Pick<CompleteHocuspocusProviderWebsocketConfiguration, 'url'>> | Required<Pick<CompleteHocuspocusProviderConfiguration, 'websocketProvider'>>);`

the `maxAttempts` property is set on `CompleteHocuspocusProviderConfiguration`. I think I need to understand what `Required` and `Pick` are doing here, but I think there is a bug in their system.

Reply:
interesting! late to the party here (sorry)... but I think the cleanup ("off" listener) function needs to reference the same function as the "on" listener. Even though the code is the same, it's not quite cleaning things up bc it's not referencing the original callback. TLDR; I think you'd need to move the callback outside into a variable, and then pass that to the "on" and the "off" (if that makes sense). We can clean this up later ofc since already merged 🤓